### ISSUE TITLE
CREST Fitting Support

### ISSUE THREAD

**Issue by blitzmann** [DarkFenX/Pyfa#383]
CCP FoxFour just release documentation on this today, see https://github.com/ccpgames/eveonline-third-party-documentation/commit/1ff6d7ac56c4b1c66b0e39155b502dfb7ec0336c

This issue will keep track of current progress.

**Comment by blitzmann** [comment]
OAUTH stuff is all new to me, so these are basically just ramblings

There are two options:

* Authentication grant: this requires the client ID and client secret key, which means we would need a dedicated server that would perform the authorization to avoid revealing the secret key. pyfa would embed a web view showing a pyfa-hosted page on some server (or even maybe simply open the web browser to display), where the user would log in via SSO. The EVE servers would then validate and confirm with the user, and redirect to a callback on the pyfa server with an authorization code.

The upside to this is that we are allowed to use a refresh token for the user, which (if I understand correctly) we can store in pyfa and make requests without re-authenticating (which would get tiresome). This is also assuming that we would only need the dedicated server to get the auth code, and won't need to keep making requests back and forth afterwards -- the complexity of this grant type will increase drastically if we need to use the dedicated server for ALL traffic between pyfa and EVE.

http://tools.ietf.org/html/rfc6749#section-4.2

* Implicit grant: does not require the client secret, and thus does not require us to use a dedicated server. The problem with this is that refresh tokens are not supported, so every 20 minutes the user would be presented with a dialog asking them to log into the SSO.

Need to test: Register custom handler with the OS that which handle opening a `pyfa://crestauth/?blah=blah` URL and test functionality.

**Comment by narthollis** [comment]
You are correct that you can store a refresh token and use it to later get a fresh access token. (Assuming the refresh token hasn't expired - which you have no idea about until you make the request)


As for how to obtain the auth code and resolve that into an access/refresh token...

http://stackoverflow.com/questions/4419915/how-to-keep-the-oauth-consumer-secret-safe-and-how-to-react-when-its-compromis
https://developers.google.com/identity/protocols/OAuth2InstalledApp

It seams the general consensus is to just keep the secret in the app, probably adding it as part of the build process and telling source users how to obtain their own secret.

Alternatively there is the proxy scenario where the client secret is kept on a 'secure' web service, but this could simply allow just about anyone to use that service to auth using your client secret anyway....

**Comment by blitzmann** [comment]
> It seams the general consensus is to just keep the secret in the app, probably adding it as part of the build process

Unfortunately this is not an option. Having the secret shipped with the client immediately makes it insecure, especially in our situation and the language we use - it's very difficult to hide things in python -- even the EVE client can be decompiled back into it's source.

> Alternatively there is the proxy scenario where the client secret is kept on a 'secure' web service, but this could simply allow just about anyone to use that service to auth using your client secret anyway....

If someone goes to a webpage specifically made for pyfa, signs into SSO, and is redirected to a webpage or application that is controlled by pyfa, no other application can really use that since that application will need to be redirected elsewhere? Hmm

**Comment by narthollis** [comment]
Proxy / Web Service:

Source | Action | Destination
------ | ------ | -----------
Pyfa | Get Token | Web Service
Web Service | Redirect (with client id) | EVE SSO
EVE SSO | Redirect (with auth code) | Web Service
Web Service | Get Token (with auth code, client id, and client secret) | CREST
Web Service | Return (with Token) | Pyfa

The problem here is that for the web service to be even equal to storing the secret in the install, the application needs to prove in some way that it is allowed to request a token.

And so you are again in the position of having to store a secret in the application, just this time its a web service secret instead.

There is no good solution to the client secret problem. You can obfuscate the secret however you want, but even a heavily obfuscated secret in a compiled C/C++ application is going to be fairly easy to retrieve.

The opening page of the Google OAuth2 documentation states it quite clearly:
> Installed apps are distributed to individual machines, and it is assumed that these apps cannot keep secrets.

It further states toward the end of the overview:
> The client ID and client secret obtained from the Developers Console are embedded in the source code of your application. In this context, the client secret is obviously not treated as a secret.

The best thing that can be done is for the Pyfa client to have access to the smallest scope the makes sense for it to have, in this case simply fitting read/write.

OAuth was never really written with desktop applications in mind, so these are the problems we run into.

**Comment by blitzmann** [comment]
Yeah, was thinking about this last night. Even with a web proxy, pyfa would have to prove its pyfa somehow, whi h can be easily spoofed. Ill talk with foxfour to see what exactly the consequences might be for someone to obtain the client secret, cause it seems to me that it doesnt really matter (especially if we can already access the data without authenticatijg the client with implict grant)

**Comment by blitzmann** [comment]
I just got confirmation from FoxFour on TweetFleet Slack that it should be okay to ship with client secret as of right now. http://pastebin.com/TfWg1ywi

**Comment by Regner** [comment]
I have asked a good friend who is a lot smarter than me to come and offer his feedback/suggestions. Posting mainly to subscribe to this issue and get an alert when he does respond.

**Comment by jimpurbrick** [comment]
OAuth 2 for native apps is fiddly. The easiest and clearest approach would be to launch a browser which opened a version of crestexplorerjs (https://github.com/jimpurbrick/crestexplorerjs) customized to display tokens to the user which can be copied in to pyfa. 

The copying and pasting can be avoided using several mechanisms discussed here http://wiki.oauth.net/w/page/27249271/OAuth%202%20for%20Native%20Apps. Allowing the redirect URL to open a web server embedded in pyfa might be pretty slick and should work on all platforms, but you'd have to have the token copying fallback in case firewalls or permissions don't allow you to run a local server.

Although it might be tempting, I would avoid embedding a browser in pyfa as a lot of the security provided by pyfa (and the old API) comes from not giving passwords to 3rd party apps, which is effectively what's happening with an embedded browser.

Authorizing an app is pretty low friction once a character is singed in to the EVE SSO, but CCP should probably consider increasing the token lifetime to cover a reasonable 3rd party app session length to avoid reauthorizing multiple times during a session.

**Comment by blitzmann** [comment]
Thanks for your input =) I've already got a server going that partially works - it extracts the token, but I haven't yet written the logic to get the access token. Shouldn't be too difficult. However I did not consider users that may be restricted by firewall settings. I'll have to give it some more thought on how to handle that situation.

What's your thought on shipping pyfa with the client key and secret vs having the user make one for themselves through the developer portal, in terms of security for both users and CCP? As I understand it, since we're talking about a single-user environment here, it should't be problematic unless the user database with their refresh token falls into malicious hands which can then use our client info with their refresh token to write to their fittings.

**Comment by jimpurbrick** [comment]
From an OAuth perspective pyfa is closer to a JavaScript client than a webserver (albeit with a more fiddly workflow for getting tokens) so I would use the Implicit Grant flow (https://tools.ietf.org/html/rfc6749#section-4.2).

While having pyfa pretend to be a webserver and using the Authorization Grant Flow (https://tools.ietf.org/html/rfc6749#section-4.1) might be tempting to get refresh tokens it would require users to sign up as a developer to get a client ID/secret, which isn't CCPs intention for the developer license and may require everyone using pyfa to have paid CCP for time/PLEX/fanfest/vegas/stuff.

If CCP can increase the lifespan of an implicit grant flow to something around the 90th percentile of pyfa session lengths (8 hours?) then you'll just need to click a single button in a browser when you start pyfa to get a token assuming you get the webserver working and people keep themselves signed in to the SSO, which isn't too onerous.

**Comment by blitzmann** [comment]
Implicit would indeed be the best way for pyfa to access the resources, however that 20 minute limit is pretty restrictive. It would be awesome if that was extended. The plan right now is to ship with our own client ID/Secret (as there's no way we could hide it), and possibly have the option for users to give their own if they wish.

I've thought of possibly giving user option to encrypt their refresh tokens (in case their database is compromised), and prompt them for decryption password when pyfa starts or the first time they attempt CREST access. But something such as this would probably not be in the initial release.

**Comment by Regner** [comment]
I have been looking at bumping that 20 minute limit to 1 hour. Would that saw you enough?

**Comment by blitzmann** [comment]
I replied to you on Slack =)

**Comment by Regner** [comment]
Going to try and keep the discussion here so if needed we can easily direct other people here.

Out of curiosity how to do imagine the user flow for this to be?

**Comment by blitzmann** [comment]
The intended flow would have been as such:

1) User opens preferences or other CREST management window and hits "Add Character" which opens browser and directs them to SSO login
2) User logs in, and is redirected to local webserver
3) Character is added to a list of available characters and is saved to the DB along with refresh token. Login once and forget.

To export, I was gonna rework the current export interface which exports to clipboard to also give user dropdown menu of saved characters to POST to CREST. And to import: https://gyazo.com/cf6b5a47ffebc40717d4ac1316e64778

This kind of flow, where the characters are saved and are automatically authenticated, would require authorization grant with refresh tokens. But as stated on Slack after having thought about it, if user database is ever compromised then they are at risk of another party modifying their fits. Which is why I was thinking about encrypting refresh token based on user-supplied password. While I think this would be easier on the user as it will be quicker than opening a port and browser, signing in, and then going back to pyfa, I'm still not sure about it.

If the limit can be bumped up to 1 hour or more, I think that will be a good start. TBH, users will probably not be getting and posting fits all that often, and when they do it will probably be after a session of theorycrafting, so logging in when needed might not be too bad.

That said, sometimes it's not as simple as opening the browser and hitting "Authorize". I have to consider those that have multiple accounts. If someone has three accounts and wants to post some fittings to each, then they have to login three different times. This is where refresh tokens would definitely come in handy.

As for managing fittings, possible future CREST features must also be considered, such as Corp fittings and modifying existing fits (I plan to link pyfa fits to their EVE fitting so that that when CREST ). 

So basically I have three options:
1) Use Implicit Grant and have user login whenever they need to GET or POST. This may become more cumbersome if additional features are added in CREST such as modifying existing fits.
2) Use Authentication Grant with pyfa's client ID/Secret and force refresh token encryption that user will unlock once during a session. Helps with fluid user interaction, and will help those with multiple accounts.
3) Allow user to supply their own client ID/Secret which allows the benefits of using the refresh tokens without the risk.

**Comment by jimpurbrick** [comment]
This would be a great plan if the web server were hosted somewhere it could keep secrets and tokens safe. Hooking up with fleet-up or a similar service might be a good idea if you don't want to run your own servers.

**Comment by narthollis** [comment]
As previously discussed, the problem with using an external service to keep the secrets safe, is how do you verify that Pyfa is allowed to access those tokens?

Also, the web server that @blitzmann is referring to in his authentication work flow is a way of accepting a redirect from EVE SSO with the auth code. This web server would be embedded into Pyfa and is the way to get around messy protocol handlers (which no user shell handles consistently) or having the user copy/paste values into Pyfa.

If keeping the secret safe is the primary concern, then the Implicit grant path is really the only option.

Otherwise, I don't think there is any greater security risk storing the refresh tokens than storing the the api key/vcode pairs. (That said, an optional encryption is not a bad idea for those people who would like to use it)

**Comment by blitzmann** [comment]
Yeah, there would be no way of authenticating to a proxy server since it will also need a secret that cannot be kept secure.

I think I will focus on using Implicit grant for default communications and allow users to put in their own client ID / Secret if they wish to do so, but I'll still keep the encryption as a possibility. Keeping our stuff secret is not a concern in and of itself being a single-user application, but since we frequently have people uploading their database for support / debugging, I can't allow their refresh tokens to be plaintext.

As for current progress, I have a very crude implementation on my repo. I hate the way I've put it together, so I'll be refining it in the coming weeks. But hey, it works!

**Comment by blitzmann** [comment]
Also, @Regner, this sounds like an interesting solution that twitter came up with: https://groups.google.com/forum/#!topic/twitter-development-talk/wYrenYbIsjk 

It's basically an official way for the user to create their own developer licence and generate their own application id / secret, which can then be sent to the callback and implemented in the application. Perhaps it might be something CCP would want to look at?

**Comment by blitzmann** [comment]
A working version has been pushed to the `crest_fitting` branch. There are two modes: implicit and user-defined client settings. 

* Under implicit, you can log in with one character at a time. This is mostly due to me being lazy and not wanting to write logic to juggle multiple characters with limited authentication time. Any CREST activity will be done under that character. The character expires after the time CCP sets (which hopefully @Regner can get raised =D). There is a timer that is set that can execute some code when the expiration is met in case we need it.
* Under user defined, user enters their own clientID and Secret. This will allow them to save all their logged in characters and select which one to use for CREST activity. This may help folks with multiple accounts, those that like to simply be able to select the character to work with, or those that simply have trust issues and want their own client info handling their data.

There's still a few refinements that I wish to make, but I need to make sure the windows and OS X skeletons have the needed modules such as BaseHTTPServer.

Going forward past the initial release I will be linking fits that are imported and exported so that if CCP ever implements an "update" fit option we already have that linkage.

Known issues:
* Changing clientID and Secret and trying to access CREST with refresh token bound to the old settings hangs the program. Still need to do some error management.
* Changing from implicit to user-supplied mode probably doesn't work until reopening pyfa
* (more of a todo): make sure that logoff timer is actually reset when changing characters under implicit mode

#### Screenshots
I'm open to suggestions on how best to display these infos.

Implicit mode, right after logging in
![implicitlogin](https://cloud.githubusercontent.com/assets/3904767/10708217/b3a91446-79ce-11e5-9540-d0712b477cb8.PNG)

Implicit mode, exporting current fit to character (the test alignment is out of whack, but I also plan to incorporate this into the existing clipboard export dialog)
![implicitexport](https://cloud.githubusercontent.com/assets/3904767/10708221/d01f01c6-79ce-11e5-90c1-692233ac9dd1.PNG)

User mode, character management:
![charbrowse](https://cloud.githubusercontent.com/assets/3904767/10708222/d823ddd8-79ce-11e5-98da-961834caa986.PNG)

User mode, browse and import fits:
![charbrowse](https://cloud.githubusercontent.com/assets/3904767/10708223/e173ab02-79ce-11e5-8e0a-716cc68efff0.PNG)

**Comment by narthollis** [comment]
Looks good.

I think the Implicit / User defined split is a good compromise to the secret problem, Nice work.

**Comment by blitzmann** [comment]
I've been working on the osx and windows skeletons, and should have working clients available for testing by tonight hopefully

More known issue:
* was able to get a "Address already in use" when running two pyfas in OS X due to both trying to listen on same port.
* The server binds to the address when pyfa starts (because it is done when the CREST service initializes). Move this out of init and only do it when startServer is called.

**Comment by cncfanatics** [comment]
The proper approach here is very likely to only start the server when
needed to get a response back ? That way two pyfas should be able to
co-exist nicely.

**Comment by blitzmann** [comment]
Yep, already done =) The server also has a 60 second window before it's shut off.

Making a pre-release now. =D

**Comment by narthollis** [comment]
In regards to fallback, I think the best option is a basic middleman page.

CREST redirects to a public Pyfa webpage. The Pyfa webpage attempts to call back to the application by JSONP (to avoid potential cross origin issues). If this call is successful the middleman page tries to close its self, or failing that displays a "All done" message to the user.

The middleman page also displays the auth code to the user, with instructions on how to paste it into Pyfa via a fall back method.

The middleman does nothing with the auth code other than pass it to the application or displays it to the user.

Being basic HTML/JS this page could be hosted with GitHub pages.

**Comment by blitzmann** [comment]
I was thinking of exactly that, but was waiting to look into hosting. I totally forgot about GitHub pages though!

**Comment by jimpurbrick** [comment]
crest explorer and contactjs are both hosted on github pages and that works
well

**Comment by blitzmann** [comment]
Merged CREST support with `singularity` branch

**Comment by Regner** [comment]
You still seem to have a lot of hard coded links for Sisi: https://github.com/DarkFenX/Pyfa/blob/2a8f3e4855ab5fdc8a25376d20dd1718ce26927d/service/crest.py#L122

https://api-sisi.testeveonline.com/ is Sisi. TQ is https://crest-tq.eveonline.com/

**Comment by Nothing4You** [comment]
Just tested this on TQ.
Created new application: ![Settings](http://i.imgur.com/HNIwPIi.png)

However, when I try to add a character (after saving the settings in pyfa) I see `https://login.eveonline.com/oauth/authorize?response_type=code&redirect_uri=http://localhost:6461&client_id=&scope=characterFittingsRead+characterFittingsWrite&state=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx` in my browser (note that the `client_id` is missing) telling me `{"error":"invalid_request", "error_description":"Some parameters are either missing or invalid"}`

**Comment by Nothing4You** [comment]
Tried again now and I got a new error with a wrong url due to the wiki being wrong (already corrected).

Now that I fixed the URL I can add a character.

**Comment by Nothing4You** [comment]
Another update: I wasn't able to do anything with the characters I added until I restarted pyfa.

**Comment by blitzmann** [comment]
Dammit, I meant to change those hardcoded links. fml

Will hopefully get a point release out tonight

**Comment by blitzmann** [comment]
https://github.com/DarkFenX/Pyfa/releases/tag/v1.16.1

Should fix the TQ CREST issues, thanks @Regner 

@Nothing4You: Yeah, switching between the modes doesn't work very well at the moment. It's fortunate that it's not needed often. I have no idea why clientID would be missing, unless it's another side effect of switching from Implicit to User defined.

**Comment by blitzmann** [comment]
I've done a bit more work on the logic of the menus - they should now correctly react to logging in, out, and switching modes.

**Comment by svetlichny** [comment]
Chiming in to add that in 1.16.2/osx/wx3 CREST integration via user-supplied details is still not working for me. 

First, if I provide credentials in settins, then go to the menu CREST > Manage Characters > Add Character, pyfa opens authentication URL in the browser without correctly inserting the Client ID from settings to the callback URL, it remains like this: 
```
https://login.eveonline.com/oauth/authorize?response_type=code&redirect_uri=http://localhost:6461&client_id=&scope=...
```
and CCP's OAuth reponds with
 ```
{"error":"invalid_request", "error_description":"Some parameters are either missing or invalid"}
```

Second, even if I fix the URL manually with specifying my ClientID param, it redirects me to the auth page where I press 'Authorize' button and it redirects me back to localhost, which the browser cannot access, because "Server not found". After trying this back and forth a couple of times pyfa's localhost server has finally responded with "all done, you may close this window", but my character did not appear in the character management window.

**Comment by noenedrops** [comment]
Chiming in to add that in 1.16.2/osx/wx3 CREST integration via user-supplied details is still not working for me. 

First, if I provide credentials in settins, then go to the menu CREST > Manage Characters > Add Character, pyfa opens authentication URL in the browser without correctly inserting the Client ID from settings to the callback URL, it remains like this: 
```
https://login.eveonline.com/oauth/authorize?response_type=code&redirect_uri=http://localhost:6461&client_id=&scope=...
```
and CCP's OAuth reponds with
 ```
{"error":"invalid_request", "error_description":"Some parameters are either missing or invalid"}
```

Second, even if I fix the URL manually with specifying my ClientID param, it redirects me to the auth page where I press 'Authorize' button and it redirects me back to localhost, which the browser cannot access, because "Server not found". After trying this back and forth a couple of times pyfa's localhost server has finally responded with "all done, you may close this window", but my character did not appear in the character management window.

**Comment by thorr18** [comment]
I get a long URL with client_id included. 
I opened 1.16.2/osx/wx3 , clicked CREST , clicked Login To Eve, Chrome browser popped up displaying Chrome saved credentials, clicked log in, clicked authorize, closed browser, exported a fit to Eve in Pyfa, and verified it appeared on my Eve character.

**Comment by blitzmann** [comment]
@thorr18, I believe @noenedrops is talking about switching to user-supplied client details. I'll have to look at the os x client again and see if I can duplicate.

@noenedrops, The server only listens about 60 seconds after the initial request, which may be why you were getting a not found error. If it says "done" but the character isn't added, it might be because the state is not correct. Run with the `-d` flag to print debug messages, which may help diagnose exactly what is going on.

**Comment by blitzmann** [comment]
@noenedrops, you are clicking the "Apply client settings" button after inserting your info, correct?

**Comment by blitzmann** [comment]
CREST is dead, long live ESI!

Closing this, all enhancements / todos / whatever revolving around ESI should be handled in separate issues as they crop up. 😄
